pydantic/dataclasses.py

```diff
@@ -264,6 +265,35 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
             **kwargs,
         )
 
+        if config_wrapper.validate_assignment:
+
+            @functools.wraps(cls.__setattr__)
+            def validated_setattr(instance: Any, field: str, value: str, /) -> None:
+                type(instance).__pydantic_validator__.validate_assignment(instance, field, value)
+
+            cls.__setattr__ = validated_setattr.__get__(None, cls)  # type: ignore
+
+            if slots and not hasattr(cls, '__setstate__'):
+                # If slots is set, `pickle` (relied on by `copy.copy()`) will use
+                # `__setattr__()` to reconstruct the dataclass. However, the custom
+                # `__setattr__()` set above relies on `validate_assignment()`, which
+                # in turn excepts all the field values to be already present on the
+                # instance, resulting in attribute errors.
+                # As such, we make use of `object.__setattr__()` instead.
+                # Note that we do so only if `__setstate__()` isn't already set (this is the
+                # case if on top of `slots`, `frozen` is used).
+
+                # Taken from `dataclasses._dataclass_get/setstate()`:
+                def _dataclass_getstate(self: Any) -> list[Any]:
+                    return [getattr(self, f.name) for f in dataclasses.fields(self)]
```

Comment:
Could we use a `dict` here instead of relying on `dataclasses.fields` being stable, which it may not be if the pickle and unpickle happen on different versions of the code?

I haven't looked into this deeply, but https://stackoverflow.com/questions/76143398/pickle-ignores-getstate-on-frozen-dataclasses-with-slots suggests it's fine.

Reply:
`dataclasses.fields` is documented so should be stable across all versions, and I don't think it will ever change even as a breaking change.

Reply:
@Viicos I was thinking about a case where a user-defined dataclass is modified and its `fields` change as a result. Then if we have a pickle from one version and unpickle on the next, the order (and length) of pickled values and dataclass fields may not match so the setstate below could fail. I haven't verified this though.

Reply:
I'm not entirely sure I follow, but I think it should be fine: the provided `__get/setstate__()` implementations are used to pickle the dataclass _instances_. If the class definition changes between pickling/unpickling, then it is expected to fail anyway